### PR TITLE
Narrow array on count() with positive-int

### PR DESCRIPTION
**Pull request by staabm**
With this PR we also narrow in case of unbound integer ranges 

separated from https://github.com/phpstan/phpstan-src/pull/3373

### CODE REVIEW

File: src/Analyser/TypeSpecifier.php [hunk 1]
```diff
@@ -1057,7 +1058,11 @@ private function turnListIntoConstantArray(FuncCall $countFuncCall, Type $type,
 				}
 
 			}
-			return $valueTypesBuilder->getArray();
+
+			$arrayType = $valueTypesBuilder->getArray();
```
Comment (staabm) [inline src/Analyser/TypeSpecifier.php:1062]: compensated the now passed in maybe unbound `IntegerRangeType`s

File: tests/PHPStan/Analyser/nsrt/count-type.php [hunk 2]
```diff
@@ -44,12 +44,12 @@ public function doFooBar(
 		if (count($arr) == $maybeZero) {
 			assertType('array', $arr);
 		} else {
-			assertType('non-empty-array', $arr);
+			assertType('array', $arr);
 		}
 		if (count($arr) === $maybeZero) {
 			assertType('array', $arr);
 		} else {
-			assertType('non-empty-array', $arr);
```
Comment (staabm) [inline tests/PHPStan/Analyser/nsrt/count-type.php:52]: I think I did a mistake here in previous PR, as the array can still be empty in the else branch

https://3v4l.org/409Jp

File: src/Analyser/TypeSpecifier.php [hunk 3]
```diff
@@ -1046,7 +1047,7 @@ private function turnListIntoConstantArray(FuncCall $countFuncCall, Type $type,
 					$offsetType = new ConstantIntegerType($i);
 					$valueTypesBuilder->setOffsetValueType($offsetType, $type->getOffsetValueType($offsetType), true);
 				}
-			} else {
+			} elseif ($type->isConstantArray()->yes()) {
```
Comment (staabm) [inline src/Analyser/TypeSpecifier.php:1053]: compensated the now passed in maybe unbound `IntegerRangeType`s

File: src/Analyser/TypeSpecifier.php [hunk 4]
```diff
@@ -2140,6 +2097,70 @@ public function resolveIdentical(Expr\BinaryOp\Identical $expr, Scope $scope, Ty
 		}
 		$rightType = $scope->getType($rightExpr);
 
+		if (
+			!$context->null()
+			&& $unwrappedLeftExpr instanceof FuncCall
+			&& count($unwrappedLeftExpr->getArgs()) >= 1
+			&& $unwrappedLeftExpr->name instanceof Name
+			&& in_array(strtolower((string) $unwrappedLeftExpr->name), ['count', 'sizeof'], true)
+			&& $rightType->isInteger()->yes()
```
Comment (staabm) [inline src/Analyser/TypeSpecifier.php:2103]: uses now a proper `Type` method instead of `&& $constantType instanceof ConstantIntegerType`

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**